### PR TITLE
Make @client_id optional, Oauth2Application mapping and REST

### LINKED ISSUE #57: Make client_id optional
Currently this is a required `Client#initialize` parameter, and is not used anywhere in the library that I can see.

It should at least be made optional, and anything added that depends on this value can fall back on `GET /oauth2/applications/@me`

### PR DESCRIPTION
**Pull request by z64**
Closes #57 

Todo:
- [ ] Reformat OAuth2 mapping style (remove tuple braces)

### CODE REVIEW

**Comment by RX14** [comment]
This should be 2 commits, containing the first 2 and last 3 commits respectively.

File: src/discordcr/client.cr [hunk 1]
```diff
@@ -35,10 +35,10 @@ module Discord
...
     )
 
-    # Creates a new bot with the given *token* and *client_id*. Both of these
-    # things can be found on a bot's application page; the token will need to be
-    # revealed using the "click to reveal" thing on the token (**not** the
-    # OAuth2 secret!)
+    # Creates a new bot with the given *token* and optionally the *client_id*.
+    # Both of these things can be found on a bot's application page; the token
+    # will need to be revealed using the "click to reveal" thing on the token
+    # (**not** theOAuth2 secret!)
```
Comment (meew0) [inline src/discordcr/client.cr]: A space is missing between `the` and `OAuth2`

File: src/discordcr/mappings/oauth2.cr [hunk 2]
```diff
@@ -0,0 +1,20 @@
+require "./converters"
+require "./user"
+
+module Discord
+  # An Oauth2 application, as registered with Discord, that can hold
+  # information about a `Client`'s associated bot user account and owner,
+  # among other Oauth2 properties.
+  struct Oauth2Application
```
Comment (meew0) [inline src/discordcr/mappings/oauth2.cr]: Shouldn't this be `OAuth2Application`? I'm not 100% familiar with the naming standards (@RX14?)

Comment (z64) [inline src/discordcr/mappings/oauth2.cr]: @meew0 You're right - it is `OAuth2Application` in Crystal's stdlib.

**Pull request merged**